### PR TITLE
tools: cluster-wide troubleshooting tool

### LINKED ISSUE #2349: clusterdebug: cluster-wide debugging tool for cilium deployments
### clusterdebugtool: cluster-wide debugging tool for cilium deployments

The tool should identify the most commonly encountered issues in cilium deployments in a Kubernetes environment. It should also provide hints to fix the issue, if applicable.

### Example use cases

* Cilium requires a specific version of the Kernel. The current kernel version is too old.
* Cilium Agent pod is crashing because it is unable to setup the KV store
* Cilium Agent pod is crashing because it is unable to contact the API server
* The kubernetes API server version is unsupported.

### PR DESCRIPTION
**Pull request by ashwinp**
- The tool identifies the most commonly observed errors.
- It performs various checks (k8s, cilium) and provides hints
to fix specific issues that it has identified.
- The ModuleCheck class implements the basic notion of a check.
It is pluggable in the sense that it can be extended to change
the default on-success and on-failure actions.
- The ModuleCheckGroup class provides a grouping construct for
grouping related ModuleChecks. The ModuleChecks in a ModuleCheckGroup are
run in a serial manner. The execution of a ModuleCheckGroup
terminates on the first ModuleCheck failure. In other words, if a
ModuleCheck fails, all subsequent ModuleChecks in the group will be
skipped.
- The initial commit contains two ModuleCheckGroups. One for Kubernetes
checks, and the other for Cilium specific checks.

Signed-off-by: Ashwin Paranjpe <ashwin@covalent.io>

**Summary of changes**:

Fixes: https://github.com/cilium/cilium/issues/2349

```release-note
The clusterdebug tool can help identify the most commonly encountered
 issues in cilium deployments. The tool currently supports Kubernetes 
 and Minikube clusters only.

 The tool performs various checks and provides hints to fix specific
  issues that it has identified.
```

### CODE REVIEW

**Review by aalemayhu — CHANGES_REQUESTED** [review]
LGTM, some minor things. Can you split the author file change into a separate commit? Would also git ignore the *.zip.

**Comment by ashwinp** [comment]
@scanf Done. Unstaged the `AUTHOR` file changes. Updated `.gitignore`.

**Comment by tgraf** [comment]
Tool looks great. We need to figure out how users should be running this, we have a couple of options:

1. Include it in the `cilium/cilium` image and provide a static binary with this tool embedded (using cpython or similar). This binary should also be released on github so non container image users can just run the plain binary.
2. Provide a separate container image which only includes this tool and starts it automatically. This could be achieved by putting this in a completely separate repo and refer to it from the docs or by including this via a special tag that is based on a separate `Dokerfile`, e.g. `cilium/cilium:cluster-debug`

@cilium/team Thoughts?

**Comment by ashwinp** [comment]
@tgraf Thanks for the review. I'd like to propose that we consider separating the packaging and delivery strategy (how will the user download and run the tool) in a separate PR. For now, we can build and upload the zip file to a publicly accessible location, which the users can download and simply run using the `python clusterdebug.zip` command. I've submitted a separate PR for the docs update here: https://github.com/cilium/cilium/pull/2356. As can be seen in that PR, there is TODO, which will be fixed after we upload the zip file. Please let me know your thoughts about this. Thanks.

File: clusterdebug-tool/clusterdebug/utils.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+logging.basicConfig(stream=sys.stdout, level=logging.DEBUG, format=FORMAT)
+if sys.stdout.isatty():
+    # running in a real terminal
+    # Color code source: http://bit.ly/2zPHiCK
+    logging.addLevelName(
+        logging.WARNING,
+        "\033[1;31m%s\033[1;0m" %
+        logging.getLevelName(
+            logging.WARNING))
+    logging.addLevelName(
```
Comment (aalemayhu) [inline clusterdebug-tool/clusterdebug/utils.py:31]: Why are you using the same color for both warning and error?

**Comment by tgraf** [comment]
> Thanks for the review. I'd like to propose that we consider separating the packaging and delivery strategy (how will the user download and run the tool) in a separate PR. For now, we can build and upload the zip file to a publicly accessible location, which the users can download and simply run using the python clusterdebug.zip command. I've submitted a separate PR for the docs update here: #2356. As can be seen in that PR, there is TODO, which will be fixed after we upload the zip file. Please let me know your thoughts about this. Thanks.

That's fine. I want to make sure we don't just merge & forget.

**Review by aalemayhu — APPROVED** [review]

**Review by tgraf — APPROVED** [review]

**Pull request merged**